### PR TITLE
Add support for local repositories (file://).

### PR DESCRIPTION
**Pull request by grilo**
##### ISSUE TYPE
- Bugfix Pull Request
##### COMPONENT NAME

maven_artifact
##### ANSIBLE VERSION

ansible 2.0.0.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
##### SUMMARY

Reduced code base in approximately 30%, (removed dead code, simplified some logic), using ansible's own error checking to catch the glaring ones (such as absence of required parameters).
Added ignore_checksum option, enabling artifact download without forcing remote checksum validation.
Added aliases to repository, username and password parameters for better consistency.
Fixed (unreported) issue where http credentials weren't being propagated correctly.
Module initialization parameters are mandatory according to the documentation but not in the code.
Add support for local .m2 repositories (allowing a simple 'mvn install' to work without requiring a full fledged repository).
##### TESTING

I've been using this in production on my company for a few months now (a few thousand downloads so far, very heterogeneous parameters), no issues found.

### CODE REVIEW

File: lib/ansible/modules/packaging/language/maven_artifact.py [hunk 1]
```diff
@@ -60,11 +62,11 @@
         required: true
     artifact_id:
         description:
-            - The maven artifactId coordinate
+            - The maven artifactId coordinate.
```
Comment (gundalow) [inline lib/ansible/modules/packaging/language/maven_artifact.py:41]: Descriptions should be full sentences,  eg with capital letter and full stop.

File: lib/ansible/modules/packaging/language/maven_artifact.py [hunk 2]
```diff
@@ -77,347 +79,259 @@
...
         default: present
-        choices: [present,absent]
+        choices: [ present, absent ]
     timeout:
         description:
             - Specifies a timeout in seconds for the connection attempt
         required: false
         default: 10
-        version_added: "2.3"
+        version_added: 2.3
```
Comment (gundalow) [inline lib/ansible/modules/packaging/language/maven_artifact.py]: version_added should be a string, as specified in the module checklist

File: lib/ansible/modules/packaging/language/maven_artifact.py [hunk 3]
```diff
@@ -77,347 +79,259 @@
...
     validate_certs:
         description:
             - If C(no), SSL certificates will not be validated. This should only be set to C(no) when no other option exists.
         required: false
         default: 'yes'
         choices: ['yes', 'no']
         version_added: "1.9.3"
+    ignore_checksum:
+        version_added: 2.3
```
Comment (gundalow) [inline lib/ansible/modules/packaging/language/maven_artifact.py]: String

File: lib/ansible/modules/packaging/language/maven_artifact.py [hunk 4]
```diff
@@ -77,347 +79,259 @@
...
+        choices: ['yes', 'no']
 '''
 
 EXAMPLES = '''
 # Download the latest version of the JUnit framework artifact from Maven Central
-- maven_artifact:
-    group_id: junit
-    artifact_id: junit
-    dest: /tmp/junit-latest.jar
+- maven_artifact: group_id=junit artifact_id=junit dest=/tmp/junit-latest.jar
```
Comment (gundalow) [inline lib/ansible/modules/packaging/language/maven_artifact.py]: Examples should be in multi-line with `key: value`

**Review by gundalow — CHANGES_REQUESTED** [review]
needs_revision

**Comment by grilo** [comment]
ready_for_review @gundalow

File: lib/ansible/modules/packaging/language/maven_artifact.py (hunk 1 above)
Comment (grilo) [reply lib/ansible/modules/packaging/language/maven_artifact.py:41]: The diff looks good, but this isn't picking up the changes properly. Is this something I can fix?

File: lib/ansible/modules/packaging/language/maven_artifact.py [hunk 5]
```diff
@@ -136,16 +136,34 @@
 
 EXAMPLES = '''
 # Download the latest version of the JUnit framework artifact from Maven Central
-- maven_artifact: group_id=junit artifact_id=junit dest=/tmp/junit-latest.jar
+- maven_artifact:
+    group_id=junit
```
Comment (gundalow) [inline lib/ansible/modules/packaging/language/maven_artifact.py]: `key: value` (with a space between), e.g.

`group_id: junit`

**Review by gundalow — CHANGES_REQUESTED** [review]
needs_revision just the examples that need work

**Comment by grilo** [comment]
ready_for_review

@gundalow Thank you for your assistance.

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
Hi,

Looks like this branch needs updating to fix some issues, please see this guide on how to do that cleanly.

http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

**Comment by grilo** [comment]
ready_for_review

**Comment by grilo** [comment]
ready_for_review

**Review by dharmabumstead — APPROVED** [review]

**Comment by paulstapleton** [comment]
Hi,
I've really wanted to be able to download from `file://` locations for some time, and have tried out your changes by copying the code locally and overwriting the existing module. I encountered a few issues which I would like to report.
I am running ansible 2.2.1.0 on CentOS 7.3.1611

Running the maven_artifact module gave me an error immediately:

```python
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_tStxC7/ansible_module_maven_artifact.py", line 374, in <module>
    main()
  File "/tmp/ansible_tStxC7/ansible_module_maven_artifact.py", line 310, in main
    parsed_url = urlparse(url_repository)
NameError: global name 'url_repository' is not defined
```

This occurs due to a problem in the main method:

```python
    try:
        parsed_url = urlparse(url_repository)
    except AttributeError as e:
        module.fail_json(msg='url parsing went wrong %s' % e)
```

where the variable `url_repository` has not been defined yet. Changing this to the following fixed the problem.
```python
        parsed_url = urlparse(module.params["url_repository"])
```

After fixing this I had issues with the checksum logic when downloading from the file system with a `file://` URL.

In the `main()` method is the following logic:

```python
# Does the local file already exist and match the remote md5?
    if client.checksum(artifact, dest):
        module.exit_json(dest=dest, state=state, version=artifact.version, changed=False)

    try:
        client.download(artifact, dest)
    except Exception as e:
        module.fail_json(msg=e.args[0])

    if ignore_checksum and not client.checksum(artifact, dest):
        module.fail_json(msg="I was able to download the artifact (%s), but the checksum doesn't match (%s)." % (artifact.url, dest))

```

Here we check the checksum once before and then again after downloading the artifact. However due to the implementation of the `FileSystemMavenClient` checksum function this always results in a failure:

```python
class FileSystemMavenClient:
...
    def checksum(self, artifact, dest):
        self.module.fail_json(msg=artifact.url)
        return True # There is nothing to check if the artifact is local
```
So downloading from the `file://` location always fails. I fixed this locally by doing a similar checksum calculation as for the `HttpMavenClient`.
```python
class FileSystemMavenClient:
...
    def download(self, artifact, destination):
        try:
            src = create_artifact_location(artifact)
            return shutil.copy2(src, destination)
        except (IOError, os.error) as why:
            self.module.fail_json(msg='Copy failed. %s' % (':'.join([artifact.group_id, artifact.artifact_id, artifact.classifier, artifact.extension, artifact.url, destination, str(why)])))

    def create_artifact_location(artifact):
      return posixpath.join(self.repo, artifact.url).replace('file:/', '').replace('//', '/')

    def checksum(self, artifact, dest):
        if not os.path.isfile(dest):
            return False
        with open(create_artifact_location(artifact), 'rb') as file:
            remote_md5 = hashlib.md5(file.read()).hexdigest()
        with open(dest, 'rb') as file:
            local_md5 = hashlib.md5(file.read()).hexdigest()
        return remote_md5 == local_md5
```

I hope you find my comments helpful.

**Comment by grilo** [comment]
@paulstapleton Thank you for your input, those were definitely glaring mistakes. The fixes and your checksum proposal have been incorporated.

File: lib/ansible/modules/packaging/language/maven_artifact.py [hunk 6]
```diff
@@ -281,7 +281,13 @@ def download(self, artifact, destination):
             self.module.fail_json(msg='Copy failed. %s' % (':'.join([artifact.group_id, artifact.artifact_id, artifact.classifier, artifact.extension, artifact.url, destination, str(why)])))
 
     def checksum(self, artifact, dest):
-        return True # There is nothing to check if the artifact is local
+        if not os.path.isfile(dest):
+            return False
+        with open(create_artifact_location(artifact), 'rb') as file:
```
Comment (paulstapleton) [inline lib/ansible/modules/packaging/language/maven_artifact.py]: Hi,
The method `create_artifact_location` doesn't actually yet. Sorry that I wasn't explicit in my comment (https://github.com/ansible/ansible/pull/19726#issuecomment-283656479).
I extracted the file path generation code from the `download` method into a new function:

```python
def create_artifact_location(artifact):
      return posixpath.join(self.repo, artifact.url).replace('file:/', '').replace('//', '/')
```

Comment (paulstapleton) [inline lib/ansible/modules/packaging/language/maven_artifact.py]: Hi,
The method `create_artifact_location` doesn't actually yet. Sorry that I wasn't explicit in my comment (https://github.com/ansible/ansible/pull/19726#issuecomment-283656479).
I extracted the file path generation code from the `download` method into a new function:

```python
def create_artifact_location(artifact):
      return posixpath.join(self.repo, artifact.url).replace('file:/', '').replace('//', '/')
```

**Comment by grilo** [comment]
@paulstapleton Fixed, thank you for having my back covered. Feel free to report any further issues.

**Comment by grilo** [comment]
@chrisisbeef @gundalow I've been using this version for quite some time in my company and we're doing several hundred deployments per day. Let me know if there are any pending issues.

File: lib/ansible/modules/packaging/language/maven_artifact.py [hunk 7]
```diff
@@ -41,59 +41,63 @@
...
+        aliases: [ username, user, uname, aws_secret_key ]
         description:
-            - The username to authenticate as to the Maven Repository. Use AWS secret key of the repository is hosted on S3
+            - The username to authenticate as to the maven Repository, Use AWS secret key of the repository is hosted on S3.
         required: false
         default: null
-        aliases: [ "aws_secret_key" ]
-    password:
+    url_password:
+        version_added: "2.3"
```
Comment (gundalow) [inline lib/ansible/modules/packaging/language/maven_artifact.py]: `"2.4"`

File: lib/ansible/modules/packaging/language/maven_artifact.py [hunk 8]
```diff
@@ -112,40 +125,46 @@
...
 import os
 import hashlib
 import sys
 import posixpath
-import urlparse
+import shutil
+import distutils.version
 from ansible.module_utils.basic import *
 from ansible.module_utils.urls import *
+import urlparse # Override urlparse from module_utils
```
Comment (gundalow) [inline lib/ansible/modules/packaging/language/maven_artifact.py]: Please import what you need rather than `*` 

https://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#python-imports

File: lib/ansible/modules/packaging/language/maven_artifact.py [hunk 9]
```diff
@@ -155,227 +174,165 @@
...
+            group_id = dict(required=True),
+            artifact_id = dict(required=True),
+            version = dict(default='latest'),
             classifier = dict(default=None),
             extension = dict(default='jar'),
-            repository_url = dict(default=None),
-            username = dict(default=None,aliases=['aws_secret_key']),
-            password = dict(default=None, no_log=True,aliases=['aws_secret_access_key']),
-            state = dict(default="present", choices=["present","absent"]), # TODO - Implement a "latest" state
+            url_repository = dict(default='http://repo1.maven.org/maven2', aliases=['repository_url', 'repository']),
```
Comment (gundalow) [inline lib/ansible/modules/packaging/language/maven_artifact.py]: Should the default be https, if so update docs

File: lib/ansible/modules/packaging/language/maven_artifact.py [hunk 10]
```diff
@@ -41,59 +41,63 @@
...
         description:
-            - The URL of the Maven Repository to download from.
+            - The URL of the maven Repository to download from.
             - Use s3://... if the repository is hosted on Amazon S3, added in version 2.2.
         required: false
         default: http://repo1.maven.org/maven2
-    username:
+    url_username:
+        version_added: "2.3"
+        aliases: [ username, user, uname, aws_secret_key ]
```
Comment (gundalow) [inline lib/ansible/modules/packaging/language/maven_artifact.py]: Are these common aliases with maven?
Generally with modules we just use `user:` and `password`

File: lib/ansible/modules/packaging/language/maven_artifact.py [hunk 11]
```diff
@@ -41,59 +41,63 @@
...
         description:
-            - The username to authenticate as to the Maven Repository. Use AWS secret key of the repository is hosted on S3
+            - The username to authenticate as to the maven Repository, Use AWS secret key of the repository is hosted on S3.
         required: false
         default: null
-        aliases: [ "aws_secret_key" ]
-    password:
+    url_password:
+        version_added: "2.3"
+        aliases: [ password, pwd, passwd, aws_secret_access_key ]
```
Comment (gundalow) [inline lib/ansible/modules/packaging/language/maven_artifact.py]: Are these common aliases with maven?
Generally with modules we just use `user:` and `password`

**Review by gundalow — CHANGES_REQUESTED** [review]

**Comment by gundalow** [comment]
Closing and re-opening the PR to rerun CI

**Comment by grilo** [comment]
ready_for_review

@gundalow Changes incorporated. Thank you for your feedback.

**Review by gundalow — APPROVED** [review]

**Comment by grilo** [comment]
@tumbl3w33d I believe the behaviour introduced with #28295 is already by default in this one. The same would apply to #27103 and #24518.

**Comment by tumbl3w33d** [comment]
First and foremost, I want to thank you for the effort you put into this PR. I see a lot of improvement for the module here, not only functionality-wise but you also did some nice refactorings and cleanups. 👍 

Now here comes the downside. This PR is quite a chunk to chew on. You wanted to extend the module by some certain functionality but did an entire overhaul of the module. While I'd also do that with my private/work code to which I have direct access, I find it quite difficult in this scope here in which it's quite hard to get follow-up bugfixes in quickly.
This module has 0 test coverage so far and you changed major parts of its logic. The change is also too big for my simple mind to evaluate just from looking at it. So the only option left for me would be intensive manual testing which means reproducing all of the different possible use cases.

Let's assume I'd find time and power to do so - how likely is it that:
* I will not forget some of the use cases
* some other (core) reviewer will spend the same effort

To sum it up - I'm really struggling with this one. On the one hand I really like most of your improvements (some are just a matter of taste). I would have preferred to get this proposed as several PRs, each one concise and focused on a certain topic. Like this I don't feel like taking exclusive responsibility for it. Maybe a core dev or @chrisisbeef could state their opinion on this.

**Comment by grilo** [comment]
It'll take me a few days, but I'm happy to do both (multiple PRs and automated tests).

Is it OK if the first one is just a big chunk of tests (no new functionality)?

**Comment by tumbl3w33d** [comment]
Nice you don't give up :)
Tests would be highly appreciated, of course! I've been thinking about adding some as well, I just didn't find the time yet to find my way into how unit/integration tests are done here (e.g., if it would be possible to spin up a dummy registry container).

In the long run I'd also like to get this module PEP8 compliant. It's listed as exception at the moment, but for future developments we should strive for it, I think.

Looking forward to your next PRs then! 👍

**Comment by dharmabumstead** [comment]
Closing due to inactivity. Please feel free to reopen and rebase if it's still applicable. Thanks.
